File: practice/algorithms/search/connected_cell_in_a_grid/connected_cell_in_a_grid.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <set>
#include <map>
#include <string>

using namespace std;

set<pair<int, int>> find_neighbors(const vector<vector<int>> & i_matrix, int i_row, int i_col, int i_m, int i_n)
{
    set<pair<int, int>> neighbors;
    for (int row = i_row - 1; row < i_row + 2; ++row)
    {
        for (int col = i_col - 1; col < i_col + 2; ++col)
        {
            if (row > -1 && row < i_m && col > -1 && col < i_n)
            {
                if ((row != i_row || col != i_col) && i_matrix[row][col] == 1)
                {
                    neighbors.insert(make_pair(row, col));
                }
            }
        }
    }
    
    return neighbors;
}

void dfs(map<pair<int, int>, set<pair<int, int>>> & i_graph, pair<int, int> i_start, set<pair<int, int>> & io_visited)
{
    io_visited.insert(i_start);
    
    set<pair<int, int>>::iterator it = i_graph[i_start].begin();
    for (; it != i_graph[i_start].end(); ++it)
    {
        if (io_visited.find(*it) == io_visited.end())
        {
            dfs(i_graph, *it, io_visited);
        }
    }
}

int main()
{
    int n, m;
    cin >> m >> n;
    
    vector<vector<int>> matrix(m, vector<int>(n));
    for (int row = 0; row < m; ++row)
    {
        for (int col = 0; col < n; ++col)
        {
            cin >> matrix[row][col];    
        }    
    }
    
    map<pair<int, int>, set<pair<int, int>>> graph;
    vector<pair<int, int>> vertices;
    
    for (int row = 0; row < m; ++row)
    {
        for (int col = 0; col < n; ++col)
        {
            if (matrix[row][col] == 1)
            {
                vertices.push_back(make_pair(row, col));
                set<pair<int, int>> neighbors = find_neighbors(matrix, row, col, m, n);
                graph[make_pair(row, col)] = neighbors;
            }
        }    
    }
    
    int reg_size = 0;
    for (int pos = 0; pos < vertices.size(); ++pos)
    {
        set<pair<int, int>> visited;
        dfs(graph, vertices[pos], visited);
        if (visited.size() > reg_size)
        {
            reg_size = visited.size();
        }
    }
    
    cout << reg_size << endl;
    
    return 0;
}

File: practice/algorithms/search/connected_cell_in_a_grid/connected_cell_in_a_grid.py
#!/usr/nim/env python

from itertools import product

def find_neigbors(i_matrix, i_row_num, i_col_num, i_m, i_n):
    neigbors = set([])
    for NEIGHBOR in product([i_row_num - 1, i_row_num, i_row_num + 1], [i_col_num - 1, i_col_num, i_col_num + 1]):
        if NEIGHBOR[0] in range(i_m) and NEIGHBOR[1] in range(i_n):
            if NEIGHBOR != (i_row_num, i_col_num) and i_matrix[NEIGHBOR[0]][NEIGHBOR[1]] == '1':
                neigbors.add(NEIGHBOR)
    return neigbors

def dfs(i_graph, i_start, visited=set()):
    visited.add(i_start)
    for VERT in i_graph[i_start]:
        if VERT not in visited:
            dfs(i_graph, VERT, visited)

m = int(raw_input())
n = int(raw_input())
matrix = []
for _ in range(m):
    matrix.append(raw_input().split())

graph = {}
vertices = set()
for ROW_NUM in range(m):
    for COL_NUM in range(n):
        if matrix[ROW_NUM][COL_NUM] == '1':
            vertices.add((ROW_NUM, COL_NUM))
            neigbors = find_neigbors(matrix, ROW_NUM, COL_NUM, m, n)
            graph[(ROW_NUM, COL_NUM)] = set(neigbors)

reg_size = 0
while vertices:
    visited = set()
    start = vertices.pop()
    dfs(graph, start, visited)
    if len(visited) > reg_size:
        reg_size = len(visited)
    vertices -= visited
print reg_size

